src/Calculator/Views/UnitConverter.xaml.cs

```diff
@@ -0,0 +1,411 @@
+            {
+                VisualStateManager.GoToState(this, UnitLoadedState.Name, false);
+            }
+            else
+            {
+                if (Model.IsCurrencyLoadingVisible)
+                {
+                    VisualStateManager.GoToState(this, UnitNotLoadedState.Name, false);
+                    StartProgressRingWithDelay();
+                }
+                else
+                {
+                    HideProgressRing();
+                    VisualStateManager.GoToState(this, !string.IsNullOrEmpty(Model.CurrencyTimestamp) ? UnitLoadedState.Name : UnitNotLoadedState.Name, true);
+                }
+            }
+        }
+
+        private void Units1_IsEnabledChanged(object sender, DependencyPropertyChangedEventArgs e)
+        {
+            if ((Units1.Visibility == Visibility.Visible) && Units1.IsEnabled)
+            {
+                SetDefaultFocus();
+            }
+        }
+
+        private void OnNetworkBehaviorChanged()
+        {
+            switch (Model.NetworkBehavior)
+            {
+                case NetworkAccessBehavior.Normal:
+                    OnNormalNetworkAccess();
+                    break;
+                case NetworkAccessBehavior.OptIn:
+                    OnOptInNetworkAccess();
+                    break;
+                case NetworkAccessBehavior.Offline:
+                    OnOfflineNetworkAccess();
+                    break;
+            }
+        }
+
+        private void OnNormalNetworkAccess()
+        {
+            CurrencyRefreshBlockControl.Visibility = Visibility.Visible;
+            OfflineBlock.Visibility = Visibility.Collapsed;
+
+            if (Model.CurrencyDataLoadFailed)
+            {
+                SetFailedToRefreshStatus();
+            }
+            else
+            {
+                SetNormalCurrencyStatus();
+            }
+        }
+
+        void OnOptInNetworkAccess()
+        {
+            CurrencyRefreshBlockControl.Visibility = Visibility.Visible;
+            OfflineBlock.Visibility = Visibility.Collapsed;
+
+            if (m_meteredConnectionOverride && Model.CurrencyDataLoadFailed)
+            {
+                SetFailedToRefreshStatus();
+            }
+            else
+            {
+                SetChargesMayApplyStatus();
+            }
+        }
+
+        void OnOfflineNetworkAccess()
+        {
+            CurrencyRefreshBlockControl.Visibility = Visibility.Collapsed;
+            OfflineBlock.Visibility = Visibility.Visible;
+        }
+
+        private void InitializeOfflineStatusTextBlock()
+        {
+            var resProvider = AppResourceProvider.GetInstance();
+            string offlineStatusHyperlinkText = resProvider.GetResourceString("OfflineStatusHyperlinkText");
+
+            // The resource string has the 'NetworkSettings' hyperlink wrapped with '%HL%'.
+            // Break the string and assign pieces appropriately.
+            const string delimiter = "%HL%";
```

Comment:
nit: we should consider using string.Split to simplify this code

Reply:
nice suggestion, we'll improve this part later after the C# work gets finished.